asllib/doc/ASLTypingReference.tex

```diff
@@ -8694,3822 +8526,7477 @@ \subsection{Comments}
+\begin{itemize}
+  \item All of the following apply (\textsc{none}):
+  \begin{itemize}
+    \item $\fields$ is $\None$;
+    \item define $\ids$ as the empty set.
+  \end{itemize}
+
+  \item All of the following apply (\textsc{some}):
+  \begin{itemize}
+    \item $\fields$ is $\langle(\vx, \subfields)\rangle$;
+    \item define $\ids$ as the union of the singleton set for $\vx$ and the union of applying $\usety$
+          to each type of a field in $\subfields$.
+  \end{itemize}
+\end{itemize}
+
+\subsection{Formally}
+\begin{mathpar}
+\inferrule[none]{}{
+  \usesubtypes(\None) \typearrow \overname{\emptyset}{\ids}
+}
+\and
+\inferrule[some]{
+  \ids \eqdef \{\vx\} \cup \bigcup_{(\Ignore, \vt) \usety(\vt)}
+}{
+  \usesubtypes(\langle(\vx, \subfields)\rangle) \typearrow \ids
+}
+\end{mathpar}
+
+\section{TypingRule.UseExpr \label{sec:TypingRule.UseExpr}}
+\hypertarget{def-useexpr}{}
+The function
+\[
+\useexpr(\overname{\expr}{\ve} \cup \langle\overname{\expr}{\ve}\rangle) \aslto \overname{\pow{\identifier}}{\ids}
+\]
+returns the set of identifiers $\ids$ on which the expression or \optional\ expression $\ve$ depends on.
+
+\subsection{Prose}
+One of the following applies:
+\begin{itemize}
+  \item All of the following apply (\textsc{none}):
+  \begin{itemize}
+    \item $\ve$ is $\None$;
+    \item define $\ids$ as the empty set.
+  \end{itemize}
+
+  \item All of the following apply (\textsc{some}):
+  \begin{itemize}
+    \item $\ve$ is $\langle\veone\rangle$;
+    \item applying $\useexpr$ to $\veone$ yields $\ids$.
+  \end{itemize}
+
+  \item All of the following apply (\textsc{e\_literal}):
+  \begin{itemize}
+    \item $\ve$ is a literal expression;
+    \item define $\ids$ as the empty set.
+  \end{itemize}
+
+  \item All of the following apply (\textsc{e\_atc}):
+  \begin{itemize}
+    \item $\ve$ is the typing assertion for expression $\ve$ and type $\tty$;
+    \item define $\ids$ as the union of applying $\useexpr$ to $\veone$ and applying $\usety$ to $\tty$.
+  \end{itemize}
+
+  \item All of the following apply (\textsc{e\_var}):
+  \begin{itemize}
+    \item $\ve$ is the variable expression for identifier $\vx$;
+    \item define $\ids$ as the singleton set for $\vx$.
+  \end{itemize}
+
+  \item All of the following apply (\textsc{e\_getarray}):
+  \begin{itemize}
+    \item $\ve$ is the \arrayaccess\ expression for base expression $\veone$ and index expression $\vetwo$;
+    \item define $\ids$ as the union of applying $\useexpr$ to $\veone$ and applying $\useexpr$ to $\vetwo$.
+  \end{itemize}
+
+  \item All of the following apply (\textsc{e\_binop}):
+  \begin{itemize}
+    \item $\ve$ is the binary operation expression over expressions $\veone$ and $\vetwo$;
+    \item define $\ids$ as the union of applying $\useexpr$ to $\veone$ and applying $\useexpr$ to $\vetwo$.
+  \end{itemize}
+
+  \item All of the following apply (\textsc{e\_unop}):
+  \begin{itemize}
+    \item $\ve$ is the unary operation expr
```

Comment:
is not one of the expression types above -> is not one of the following expression types